### DISCUSSION TITLE
Integrate Prometheus Exporter

### DISCUSSION THREAD

**Discussion by mrueg** [nextcloud/all-in-one#2513] [Ideas] [answered]
It would be great to have some visibility into if nextcloud is working well.
There is a prometheus exporter available here, that could be helpful.
https://github.com/xperimental/nextcloud-exporter

It's also integrated in nextcloud/helm: https://github.com/nextcloud/helm/blob/main/charts/nextcloud/templates/metrics-deployment.yaml

**Comment by szaimen** [comment]
Hi, I fear this does not fulfill all requirements mentioned in https://github.com/nextcloud/all-in-one#requirements-for-integrating-new-containers

However you might be interested in this: https://github.com/nextcloud/all-in-one/discussions/1581

**Comment by szaimen** [comment]
Hi, I just wanted to mention that we have this now: https://github.com/nextcloud/all-in-one/tree/main/community-containers#how-to-add-containers. So the community could potentially add prometheus-exporter as additional container. Feel free to ping me if you should need help on this!

**Reply by Fuseteam** [reply]
surprised nobody added one yet, is there an discussion on how to get them to work along side? I have been having issues finding much information so far

**Reply by szaimen** [reply]
I've added this to https://github.com/nextcloud/all-in-one/issues/5251 now :)

**Reply by Fuseteam** [reply]
thanks <3

**Comment by Grotax** [comment]
Hey @szaimen I started working on this but I have one issue.

The nextcloud exporter needs a token (or username and password from admin) to access the server info.
The token can be set via occ.

https://github.com/Grotax/all-in-one/blob/7eb14d442c168739eeb1044585236f3458db0e73/community-containers/nextcloud-exporter/nextcloud-exporter.json#L21C1-L36C2

But for this to work I need to add the secret variable also to the nextcloud container.

https://github.com/Grotax/all-in-one/blob/7eb14d442c168739eeb1044585236f3458db0e73/php/containers.json#L151C1-L176C35

And I don't think that is optimal for a community container.
Is there an alternative that I overlooked to get a secret and supply it to the nextcloud occ command?

**Reply by Grotax** [reply]
Ah I just saw https://github.com/nextcloud/all-in-one/discussions/3889

And the issue with the facerecognition, yea so the prometheus exporter would also benefit from a solution.
I don't want to add it that way :)

**Comment by szaimen** [comment, marked as answer]
PR is in https://github.com/nextcloud/all-in-one/pull/6837

**Reply by szaimen** [reply]
This is now released with v11.8.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by Fuseteam** [reply]
finally got around to looking into enabling it, but it seems setup is limited to either  external network access or local server access.

In our setup our nginx server, prometheus server and aio server as all separated, so we'd need some way to have internal network access in a similar way we have local_ip:8080 for container management

**Reply by Grotax** [reply]
Hi, the concern with that was that the exporter itself does not have any form of authentication.
That is why it was limited to localhost to be exposed by a proxy that adds authentication.

**Reply by Fuseteam** [reply]
I understand that, but when the network is walled off with vlans that is less of an concern in our case. Would it be an idea to add authentication to the exporter itself? or maybe it could be bound to a specific local ip rather than 0.0.0.0? or perhaps it could be limited to be accessible from a specific subnet?

In our case we don't want to expose it to metrics.nc.domain but we also don't want to run prometheus on the nextcloud server either

**Reply by Fuseteam** [reply]
actually @Grotax now i'm thinking back on this, is the concerns surrounding authentication valid for a local_ip? You'd need to on the actual network of the server to access it unauthenticated :thinking: 

All i really need is a way to access the exporter via a private ip so either 192.168.x.x, 10.x.x.x or 172.16.x.x.x to 172.31.x.x

**Reply by Grotax** [reply]
Probably not but as far as I know the code of aio does not allow more than the current solution.

Edit: I actually think it heavily depends on your environment if you can trust a local network IP range or not. 

I mean it's just a Json file that says take that image an configure it like this...

But I bet PR are welcome if you have an idea how to extend the system and then I can adjust the config for the exporter.

**Reply by Fuseteam** [reply]
i think it's basically on a way to change this i suppose :thinking: 
https://github.com/nextcloud/all-in-one/blob/bb4790ed3aa476f48f18729924cd0c0370ff34bb/community-containers/nextcloud-exporter/nextcloud-exporter.json#L13

....oh hey sneaky edit :joy: maybe i just need to setup a proxy on the server to proxy from local-ip to 127.0.0.1 :thinking:

**Reply by Fuseteam** [reply]
hmmm i tried adding the container to my instance anyway
i did this by
- stopping my containers
- selecting the exporter
- clicking save changes
- starting my containers

at the last step my containers did not want to start, it did give a error message but it was difficult for me to parse— i have unfortunately not been able to gather the logs properly, i may try another day

**Reply by Fuseteam** [reply]
found it, it was failing because i already had an exporter running at 9205, once i took that down it came up fine, i managed to put an nginx proxy in front and see the expected output when accessing through the port configured on the proxy

**Reply by Fuseteam** [reply]
@Grotax one last question, it seems that total installed apps and available app updates are disabled by default, is there a way to enable them? the upstream [readme](https://github.com/xperimental/nextcloud-exporter/blob/master/README.md) mentions flags, env variables and a config file as a option to enable them— i'm unsure if one of these are usable if the community container

**Reply by Grotax** [reply]
No it's currently not possible to adjust the config. If you want that I would suggest to do a custom deployment of the container with your own setup, I think that is the only way. If you use the integrated exporter then you have to live with the default settings.

**Reply by Fuseteam** [reply]
fair enough, given AIO handles app updates, i think it's not that crucial to have

but i suppose we ought to update the readme of the container? right now it suggests it does monitor these two metrics

**Reply by Fuseteam** [reply]
filed a PR for it: https://github.com/nextcloud/all-in-one/pull/7900
